src/modeling_library/switch/regenerate.jl

```diff
@@ -0,0 +1,82 @@
+mutable struct SwitchRegenerateState{T}
+    weight::Float64
+    score::Float64
+    noise::Float64
+    prev_trace::Trace
+    trace::Trace
+    index::Int
+    retdiff::Diff
+    SwitchRegenerateState{T}(weight::Float64, score::Float64, noise::Float64, prev_trace::Trace) where T = new{T}(weight, score, noise, prev_trace)
+end
+
+@inline regenerate_recurse_merge(prev_choices::ChoiceMap, selection::EmptySelection) = prev_choices
+@inline regenerate_recurse_merge(prev_choices::ChoiceMap, selection::AllSelection) = choicemap()
+function regenerate_recurse_merge(prev_choices::ChoiceMap, selection::Selection)
+    prev_choice_value_iterator = get_values_shallow(prev_choices)
+    prev_choice_submap_iterator = get_submaps_shallow(prev_choices)
+    new_choices = choicemap()
+    for (key, value) in prev_choice_value_iterator
+        in(key, selection) && continue
+        set_value!(new_choices, key, value)
+    end
+    for (key, node1) in prev_choice_submap_iterator
+        if in(key, selection)
+            subsel = getindex(selection, key)
+            node = regenerate_recurse_merge(node1, subsel)
+            set_submap!(new_choices, key, node)
+        else
+            set_submap!(new_choices, key, node1)
+        end
+    end
+    return new_choices
+end
+
+function process!(gen_fn::Switch{C, N, K, T},
+                  index::Int,
+                  index_argdiff::UnknownChange,
```

Comment:
It might be better to use `Diff` instead of `UnknownChange` here. There might be other `Diff` types that could be passed that are intermediate between `UnknownChange` and `NoChange` (e.g. there is an `IntDiff` already, which tracks the arithmetic difference between two integers). This method should apply to anything that's not a `NoChange`.